lib/workers.js

```diff
@@ -0,0 +1,222 @@
+};
+
+class WorkerObject {
+  constructor(options, tests, testRoot, workerIndex) {
+    this.options = options || {};
+    this.tests = tests;
+    this.testRoot = testRoot || getTestRoot();
+    this.workerIndex = workerIndex;
+  }
+
+  addConfig(config) {
+    this.options.config = JSON.stringify(config);
+  }
+
+  addTests(tests) {
+    this.tests = tests;
+  }
+}
+
+class Workers extends EventEmitter {
+  /**
+   * @param {Number} numberOfWorkers
+   * @param {Boolean} runAsSuite
+   */
+  constructor(numberOfWorkers, runAsSuite, configPath) {
+    super();
+    this.codecept = initializeCodecept(configPath);
+    this.finishedTests = [];
+    this.errors = [];
+    this.numberOfWorkers = 0;
+    this.closedWorkers = 0;
+    this.workers = [];
+    this.stats = {
+      passes: 0,
+      failures: 0,
+      tests: 0,
+    };
+
+    if (typeof numberOfWorkers === 'number' && numberOfWorkers > 0) {
+      this.numberOfWorkers = numberOfWorkers;
+      this.testGroups = runAsSuite ? this.createGroupsOfSuites(numberOfWorkers) : this.createGroupsOfTests(numberOfWorkers);
+      this.workers = createWorkerObjects(this.testGroups, this.codecept.config);
+    }
+  }
+
+  spawn() {
+    const worker = new WorkerObject({}, [], getTestRoot(), this.numberOfWorkers);
+    this.workers.push(worker);
+    this.numberOfWorkers += 1;
+    return worker;
+  }
+
+  createGroupsOfTests(numberOfWorkers) {
+    const files = this.codecept.testFiles;
+    const mocha = Container.mocha();
+    mocha.files = files;
+    mocha.loadFiles();
+    const groups = populateGroups(numberOfWorkers);
+    let groupCounter = 0;
+
+    mocha.suite.eachTest((test) => {
+      const i = groupCounter % groups.length;
+      if (test) {
+        const { id } = test;
+        groups[i].push(id);
+        groupCounter++;
+      }
+    });
+    return groups;
+  }
+
+  createGroupsOfSuites(numberOfWorkers) {
+    const files = this.codecept.testFiles;
+    const mocha = Container.mocha();
+    mocha.files = files;
+    mocha.loadFiles();
+    const groups = populateGroups(numberOfWorkers);
+
+    mocha.suite.suites.forEach((suite) => {
+      const i = indexOfSmallestElement(groups);
+      suite.tests.forEach((test) => {
+        if (test) {
+          const { id } = test;
+          groups[i].push(id);
+        }
+      });
+    });
+    return groups;
+  }
+
+  run() {
+    this.codecept.runBootstrap((err) => {
+      if (err) throw new Error(`Error while running bootstrap file :${err}`);
+      for (const worker of this.workers) {
+        const workerThread = createWorker(worker);
+        this.listenWorkerEvents(workerThread);
+      }
+    });
+    return this;
+  }
+
+  isFailed() {
+    return this.stats.failures || this.errors.length;
+  }
+
+  listenWorkerEvents(worker) {
+    worker.on('message', (message) => {
+      output.process(message.workerIndex);
+
+      switch (message.event) {
+        case event.test.failed: output.test.failed(repackTest(message.data));
```

Comment:
there should be a way to hook custom listeners to this events. Right now it's only output...
please use `this.dsipatch` to resend this message so we could use it like this:

```js
workers.on(event.test.failed, () => {
 // custom action
});
```